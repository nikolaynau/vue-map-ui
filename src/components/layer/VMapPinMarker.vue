<script lang="ts">
import { defineComponent } from 'vue';

export default defineComponent({
  inheritAttrs: false
});
</script>

<script setup lang="ts">
import { useSlots } from 'vue';
import type { DivIcon, Marker } from 'leaflet';
import type { AddPrefix } from '../../utils/types';
import { useSplitAttrs, useRef } from '../../composables';
import {
  default as VMapMarker,
  type Attrs as MarkerAttrs,
  type Props as MarkerProps
} from './VMapMarker.vue';
import {
  default as VMapPinIcon,
  type Attrs as IconAttrs,
  type Props as IconProps
} from './VMapPinIcon.vue';

export type Attrs = MarkerProps &
  MarkerAttrs &
  AddPrefix<IconProps, 'icon'> &
  AddPrefix<IconAttrs, 'icon'>;

const { default: markerAttrs, icon: iconAttrs } = useSplitAttrs(['icon']);
const slots = useSlots() as { default: unknown };

const { templateRef: markerRef, value: marker } = useRef<
  InstanceType<typeof VMapMarker>,
  Marker | null
>(obj => obj.marker);

const { templateRef: iconRef, value: icon } = useRef<
  InstanceType<typeof VMapPinIcon>,
  DivIcon | null
>(obj => obj.icon);

defineExpose({
  marker,
  icon
});
</script>

<template>
  <VMapMarker ref="markerRef" v-bind="markerAttrs">
    <VMapPinIcon ref="iconRef" v-bind="iconAttrs">
      <template v-if="slots.default" #default>
        <slot></slot>
      </template>
    </VMapPinIcon>
  </VMapMarker>
</template>
