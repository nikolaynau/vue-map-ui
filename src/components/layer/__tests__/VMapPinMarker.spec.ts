import { describe, it, expect } from 'vitest';
import { ref, reactive, h, defineComponent, onMounted, nextTick } from 'vue';
import { DivIcon, Marker } from 'leaflet';
import { mount } from '../../../../.test';
import VMapPinMarker from '../VMapPinMarker.vue';

describe('VMapPinMarker', () => {
  it('should be expose instances', () => {
    expect.assertions(2);

    const Component = defineComponent({
      setup() {
        const marker = ref<InstanceType<typeof VMapPinMarker> | null>(null);

        onMounted(() => {
          expect(marker.value?.icon).toBeInstanceOf(DivIcon);
          expect(marker.value?.marker).toBeInstanceOf(Marker);
        });

        return () => h(VMapPinMarker, { latlng: [0, 0], ref: marker });
      }
    });

    mount(Component);
  });

  it('should be reactive attrs', async () => {
    const props = reactive<any>({
      latlng: [0, 0],
      iconColor: '#000'
    });

    const marker = ref<InstanceType<typeof VMapPinMarker> | null>(null);

    const Component = defineComponent({
      setup() {
        return () => h(VMapPinMarker, { ref: marker, ...props });
      }
    });

    mount(Component);

    await nextTick();

    expect(marker.value?.marker?.getLatLng()).toEqual({ lat: 0, lng: 0 });
    expect(marker.value?.icon?.options.html).toMatchSnapshot('init html');

    props.latlng = [1, 2];
    props.iconColor = '#fff';
    await nextTick();

    expect(marker.value?.marker?.getLatLng()).toEqual({ lat: 1, lng: 2 });
    expect(marker.value?.icon?.options.html).toMatchSnapshot('change html');
  });

  it('should be render default slot', async () => {
    const marker = ref<InstanceType<typeof VMapPinMarker> | null>(null);

    const Component = defineComponent({
      setup() {
        return () =>
          h(VMapPinMarker, { ref: marker, latlng: [0, 0] }, () =>
            h('i', { class: 'icon' })
          );
      }
    });

    mount(Component);

    await nextTick();
    expect(marker.value?.icon?.options.html).toMatchSnapshot();
  });
});
