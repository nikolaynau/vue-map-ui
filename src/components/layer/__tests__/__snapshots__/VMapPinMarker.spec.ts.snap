// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VMapPinMarker > should be reactive attrs > change html 1`] = `"<div class="v-map-pin-icon__shadow"></div><div class="v-map-pin-icon__pin" style="color: rgb(255, 255, 255);"></div><!--v-if-->"`;

exports[`VMapPinMarker > should be reactive attrs > init html 1`] = `"<div class="v-map-pin-icon__shadow"></div><div class="v-map-pin-icon__pin" style="color: rgb(0, 0, 0);"></div><!--v-if-->"`;

exports[`VMapPinMarker > should be render default slot 1`] = `"<div class="v-map-pin-icon__shadow"></div><div class="v-map-pin-icon__pin"></div><div class="v-map-pin-icon__icon"><i class="icon"></i></div>"`;
